src/Dictionaries/FlatDictionary.cpp

```diff
@@ -151,6 +152,114 @@ ColumnPtr FlatDictionary::getColumn(
     return result;
 }
 
+ColumnPtr FlatDictionary::getColumnOrDefaultShortCircuit(
```

Comment:
I see that for all dictionaries the implementation of `getColumnOrDefaultShortCircuit` differs from usual `getColumns` only on several lines that are responsible for working with default values. It will be good if we can avoid code duplication and implement both `getColumnOrDefaultShortCircuit` and `getColumns` using the same function like `getColumnsImpl` that will decide what to do with defaults using arguments.
But maybe there is a reason to separate logic, wdyt?

Reply:
We can use a bool flag in the template to tell whether it is short circuit or not, but what do you think would be a suitable approach to abstract away the two different parameters `const ColumnPtr & default_values_column` and `IColumn::Filter & default_mask`? I can only think of using two `optional`, but it violates the contracts if two `nullopt` are given.

Reply:
I come up with an idea using `variant`, sth like [this](https://www.onlinegdb.com/39O17LyPg), which would be better than two parameters I think.

Reply:
done with variant, merged the short circuit logic into getColumn and getColumns, without affecting existing code.